dbuf-lsp/src/action_handler/rename.rs

```diff
@@ -0,0 +1,147 @@
+            branch_id: _,
+            name: _,
+        } => false,
+        Symbol::Constructor(_) => false,
+        Symbol::None => false,
+    }
+}
+
+/// Check if symbol can be renamed to new_name without conflicts.
+pub fn renameable_to_symbol(symbol: &Symbol, new_name: &String, ast: &ElaboratedAst) -> Result<()> {
+    if new_name.is_empty() {
+        return rename_errors::rename_to_empty_error();
+    }
+    if dbuf_language::get_bultin_types().contains(new_name) {
+        return rename_errors::rename_to_builtin_type_error();
+    }
+    if dbuf_language::get_keywords().contains(new_name) {
+        return rename_errors::rename_to_keyword_error();
+    }
+
+    match symbol {
+        Symbol::Type(t) => {
+            if dbuf_language::get_bultin_types().contains(t) {
+                return rename_errors::rename_of_buildin_type_error();
+            }
+            if !dbuf_language::is_correct_type_name(new_name) {
+                return rename_errors::rename_to_bad_type_error(new_name);
+            }
+            if t == new_name {
+                return rename_errors::rename_to_old_error();
+            }
+            if ast.has_type_or_constructor(new_name) {
+                return rename_errors::rename_to_existing_type_error(new_name);
+            }
+        }
+        Symbol::Dependency {
+            t: type_name,
+            dependency: d,
+        } => {
+            if d == new_name {
+                return rename_errors::rename_to_old_error();
+            }
+            if !dbuf_language::is_correct_dependency_name(new_name) {
+                return rename_errors::rename_to_bad_dependency_error(new_name);
+            }
+            if !type_dependency_valid_rename(ast, type_name, new_name) {
+                return rename_errors::rename_to_existing_resource_error(type_name, new_name);
+            }
+        }
+        Symbol::Field {
+            constructor: ctr,
+            field: f,
+        } => {
+            if f == new_name {
+                return rename_errors::rename_to_old_error();
+            }
+            if !dbuf_language::is_correct_field_name(new_name) {
+                return rename_errors::rename_to_bad_field_error(new_name);
+            }
+            let t = ast.get_constructor_type(ctr).expect("valid symbol");
+            if !constructor_field_valid_rename(ast, t, ctr, new_name) {
+                return rename_errors::rename_to_existing_resource_error(t, new_name);
+            }
+        }
+        Symbol::Alias {
+            t: _,
+            branch_id: _,
+            name: _,
+        } => return rename_errors::rename_of_alias_error(),
+        Symbol::Constructor(_) => return rename_errors::rename_of_constructor_error(),
+        Symbol::None => return rename_errors::rename_none_symbol_error(),
+    };
+
+    Ok(())
+}
+
+/// Check if any type's dependency can be renamed to rename.
+fn type_dependency_valid_rename(ast: &ElaboratedAst, type_name: &str, rename: &str) -> bool {
```

Comment:
I'd expand on what "`valid_rename`" means here -- that it's the name _not_ used as a dependency or as a field in a type.
Maybe I'd then flip the bool and test whether it _is_ used as a dependency or a field.
Maybe I'd then introduce some term for a thing that is either a dependency or a field of a type for brevity.